add tests for ndarray

mattip opened the pull request:
Strangely, this test passes. It seems the argument name in a cdef getter is not important, since the name in [`ndarray.size`](https://github.com/cython/cython/blob/3.0a1/Cython/Includes/numpy/__init__.pxd#L265) is `ndarray`, not `self`.

mattip commented:
Tests passed without changing the paramter name in ed3b972. In 51f015d I changed it anyway. I don't know if the fact ed3b972 passes should be considered a bug or not.

scoder commented:
> don't know if the fact ed3b972 passes should be considered a bug or not.

Definitely a bug. It means that the code doesn't do what it's written for. Not even sure if it's used at all then…

da-woods commented:
> Not even sure if it's used at all then…

It's at least using the correct bit of code. It looks like `@property` on a `cdef` function is a slightly odd case that's probably quite limited in what it supports so maybe misses some checks. The generated code comes from `AttributeNode` https://github.com/cython/cython/blob/caa0f7f445728c386c50d9b88ce5f043ab224c3b/Cython/Compiler/ExprNodes.py#L7146-L7148 where it just copies the `cname` with no reference to the argument name.

Probing `entry.pos` on that confirms that it comes from @mattip's __init__.pxd file.

The `Entry` is initialized from a `CVarDefNode` which is presumably generated from property, but that's where I get lost.

I'm almost certain I've seen this bug before but I can find it right now. But we can at least be confident that the code comes from the right place.

scoder commented:
Hmm, I had completely missed what this does:
https://github.com/cython/cython/blob/66e47e8daeb38877b7d3a137fd574f3326e067b9/Cython/Compiler/ParseTreeTransforms.py#L2274-L2275
It expects the property function's body to consist of exactly one simple function call, and then extracts the name of that function and calls it with one argument. 8-O

That's the wrong thing to do. Instead, it should generate the actual property "method" and call that.

mattip commented:
> it should generate the actual property "method" and call that.

Since this is a getter decorator, is it wrong to think the c function to call will be a getter function with the same signature?

scoder commented:
> Since this is a getter decorator, is it wrong to think the c function to call will be a getter function with the same signature?

I don't see any reason why we should impose this assumption on user code.

IMHO, we should treat extern property methods like inline functions in .pxd files. Their interface is well defined, and then they can do whatever they want.

mattip commented:
> it should generate the actual property "method" and call that.

Is there an example for regenerating a node and entry like this for a c-level decorator? I am afraid my cython-foo is not up to this task.

scoder commented:
I think we need to remove this:
https://github.com/cython/cython/blob/76d8bf7990fd9622520b7c41408478bb03d19be7/Cython/Compiler/Nodes.py#L1733-L1735
and then simply call the function that it generates afterwards.

mattip commented:
In 5ec5160 I removed the offending assignment and adopted the suggestion to remove skipping `generate_function_definition`. It took me a while to understand why that was the correct thing to do. I also added a test for an arbitrary getter function, which before would have failed. There is now a new warning about unused function declarations.

I started adding a test about the misnamed argument but could not get it to properly fail. If this passes CI I will add the non-failing test, maybe someone has an idea.

scoder commented on Cython/Compiler/Nodes.py:
```diff
@@ -1811,7 +1811,10 @@ def generate_function_definitions(self, env, code):
 
         # ----- GIL acquisition
-        acquire_gil = self.acquire_gil
+        if self.entry.is_cgetter:
+            acquire_gil = False
+        else:
+            acquire_gil = self.acquire_gil
```
The getter function might still be defined with a `with gil` declaration. Is there anything specific that didn't work if you leave this out?

scoder commented:
Looks like only the `numpy_parallel` test broke from this. Probably a GIL issue:
`Fatal Python error: PyThreadState_Get: no current thread`

The NumPy `ndarray` properties are currently declared as requiring the GIL, as are the macros. Is that correct? If so, then Cython might be generating refnanny code in them. If the prange code tries to call them without holding the GIL (which it shouldn't, if they require the GIL, but I'm making a pure guess here), then this can lead to the crash.

What's the right fix here? Require the GIL and fix the parallel code if it is incorrect? Or declare the properties as nogil? (We also might have to integrate the properties into the GIL validation, BTW.)

scoder commented:
@mattip is it correct that the macros require the GIL? Their current implementation in NumPy doesn't, but I can't find anything in the NumPy API docs that would suggest or even guarantee that.

mattip commented:
I don't think they should require the GIL. The change in the code is in this snippet:
```
for i in prange(x.shape[0], nogil=True):
    x[i] = i - 5
```

where in the current implementation (before this PR) `x.shape[0]` became `PyArray_DIMS(__pyx_v_self)[0]`, but now (after this PR) is 
```
static npy_intp *__pyx_f_5numpy_7ndarray_shape(PyArrayObject *__pyx_v_self) {
  npy_intp *__pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("shape", 0);
  __pyx_r = PyArray_DIMS(__pyx_v_self);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}
```

In pre-3.0 (for comparison) `x.shape[0]` became `__pyx_v_x->dimensions[0]`

Is there a way to tell cython not to add `RefNanny` for c_getter code?

da-woods commented:
@scoder You can find them used internally in Numpy in places where the GIL has been released. As an example:

https://github.com/numpy/numpy/blob/dfd4c9de8d849363e3d4cda7a0792ed0670165cf/numpy/core/src/multiarray/calculation.c#L128-L135

`PyArray_SIZE` and `PyArray_DATA` are both used inside the no-gil block.

@mattip I think but I'm not 100% sure that `cdef npy_intp size(self) nogil:` should make the refnanny code disappear.

mattip commented:
@da-woods thanks for the hint. Any ideas why this node has no `acquire_gil` attribute in the first place? That was the reason I added the clause on line 1814.

scoder commented:
I've taken the liberty to rewrite the cgetter implementation in #3571.

scoder commented:
Thanks @mattip for helping me understand how the things should work here, even if I ended up removing a large part of your original implementation. It was very helpful to have a good set of tests and seeing where I needed to put my fingers.

mattip commented:
Should I refactor this PR to contain only the additional test?

scoder commented:
> Should I refactor/rebase this PR to contain only the additional test?

Ah, yes, the `numpy_attributes` test would be nice to have. I already stole your other one.

Travis seems to be happy with my implementation so far, so I'll merge it as soon as the CI build is done, and then you can add your test in this PR (or another, as you wish).

scoder commented:
> `PyArray_SIZE` and `PyArray_DATA` are both used inside the no-gil block.

Speaking of `PyArray_DATA`, we don't have a property for that currently. @mattip that because the type of the `data` field is `char*` but the function version returns a `void*`?

I mean, Cython 3.0 could be the right moment to change the type here…

mattip commented:
> Speaking of PyArray_DATA, we don't have a property for that currently

There is code in the wild that expects a `char *`. We could use `PyArray_BYTES`. I think I didn't do it since 
- there was some talk of trying to make that a read-only attribute, which did not go anywhere
- as far as I can tell there is no test

So in short: we should probably convert it to a `@property`, and remove the `*base` field as well since it is not public.

scoder commented:
I didn't entirely mean to close this just yet.

Let's keep `*base* in there, but I added a comment at least. I would guess that it's not used much, but I don't think Cython should decide when to break such code. :)

da-woods commented:
Can't `base` be converted to a property too https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_BASE? Probably staying as a `PyObject*` return type since it returns a borrowed reference.

It might break code that tries to set the base though - there is a setter function but presumably `cdef` properties don't support setters.

scoder commented:
> It might break code that tries to set the base though

As I said, I don't think it's up to the Cython project to decide when to break such code. As a bit of background, the main reason why this feature was built was that [we found](https://github.com/numpy/numpy/issues/11803#issuecomment-417837512) the NumPy project to be a much better place to maintain the `numpy/*.pxd` API declarations. But to do that, they needed a way to abstract from the internal fields of the `ndarray` object struct. Thus, C properties.

BTW, there's also #2663 regarding the "to GIL or not to GIL" question.

@mattip do you think it's a good time now to officially hand over the maintainership, even if we keep shipping a copy for the time being? I created #3573 to discuss the further details.

> There is a setter function but presumably cdef properties don't support setters.

That's a current limitation, yes. Not a technical one, this could be implemented, but I considered getters a much stronger use case than setters for now.

scoder commented:
> I considered getters a much stronger use case than setters for now.

OTOH – if NumPy takes over the maintenance, then it would be bad if a specific Cython version rejected their syntax… Let's discuss that in #3573.
